tests/e2e/improved_csi_idempotency.go

```diff
@@ -0,0 +1,572 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	fnodes "k8s.io/kubernetes/test/e2e/framework/node"
+	fpod "k8s.io/kubernetes/test/e2e/framework/pod"
+	fpv "k8s.io/kubernetes/test/e2e/framework/pv"
+)
+
+var _ = ginkgo.Describe("[csi-block-vanilla] [csi-file-vanilla] "+
+	"[csi-guest] [csi-supervisor] Improved CSI Idempotency Tests", func() {
+	f := framework.NewDefaultFramework("idempotency-csi")
+	const defaultVolumeOpsScale = 30
+	const defaultVolumeOpsScaleWCP = 29
+	var (
+		client            clientset.Interface
+		fullSyncWaitTime  int
+		namespace         string
+		scParameters      map[string]string
+		storagePolicyName string
+		volumeOpsScale    int
+		isServiceStopped  bool
+		serviceName       string
+	)
+
+	ginkgo.BeforeEach(func() {
+		bootstrap()
+		client = f.ClientSet
+		namespace = getNamespaceToRunTests(f)
+		scParameters = make(map[string]string)
+		isServiceStopped = false
+		storagePolicyName = GetAndExpectStringEnvVar(envStoragePolicyNameForSharedDatastores)
+		nodeList, err := fnodes.GetReadySchedulableNodes(f.ClientSet)
+
+		framework.ExpectNoError(err, "Unable to find ready and schedulable Node")
+
+		if !(len(nodeList.Items) > 0) {
+			framework.Failf("Unable to find ready and schedulable Node")
+		}
+
+		if guestCluster {
+			svcClient, svNamespace := getSvcClientAndNamespace()
+			setResourceQuota(svcClient, svNamespace, rqLimit)
+		}
+
+		if os.Getenv("VOLUME_OPS_SCALE") != "" {
+			volumeOpsScale, err = strconv.Atoi(os.Getenv(envVolumeOperationsScale))
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			framework.Logf("VOLUME_OPS_SCALE is set to %v", volumeOpsScale)
+		} else {
+			if vanillaCluster {
+				volumeOpsScale = defaultVolumeOpsScale
+			} else {
+				volumeOpsScale = defaultVolumeOpsScaleWCP
+			}
+		}
+		framework.Logf("VOLUME_OPS_SCALE is set to %v", volumeOpsScale)
+
+		if os.Getenv(envFullSyncWaitTime) != "" {
+			fullSyncWaitTime, err = strconv.Atoi(os.Getenv(envFullSyncWaitTime))
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			// Full sync interval can be 1 min at minimum so full sync wait time has to be more than 120s
+			if fullSyncWaitTime < 120 || fullSyncWaitTime > defaultFullSyncWaitTime {
+				framework.Failf("The FullSync Wait time %v is not set correctly", fullSyncWaitTime)
+			}
```

Comment:
should we fail here? or set it to a proper value for the tests and continue...

Reply:
it is a fail-fast here if the desired value is not set  as per the readme document - anything against the readme is okay to fail
 
Also, If you are suggesting to modify the deployment file on the setup if the fullsyncwaittime is not set - and check the csi pods restarts with those values ?

This will require an additional PR to address comment - let me know what you think

Reply:
In few other tests it is set to `defaultFullSyncWaitTime` if proper value was not passed.

Reply:
it is already set to defaultFullSyncWaitTime in line 95 if value is not set